crates/bin/prove_block/src/state_utils.rs

```diff
@@ -42,52 +37,55 @@ pub(crate) async fn get_processed_state_update(
+    let compiled_classes = build_compiled_class_and_maybe_update_class_hash_to_compiled_class_hash(
+        provider,
+        block_id,
+        &accessed_addresses,
+        &address_to_class_hash,
+        &mut class_hash_to_compiled_class_hash,
+    )
+    .await?;
 
-    // Collect keys without consuming the HashMaps by borrowing and cloning the keys
-    let accessed_addresses: HashSet<Felt252> = address_to_class_hash
-        .keys()
-        .copied()
-        .chain(address_to_nonce.keys().cloned())
-        .chain(storage_updates.keys().cloned())
-        .collect();
-
-    FormattedStateUpdate {
-        address_to_class_hash,
-        address_to_nonce,
-        class_hash_to_compiled_class_hash,
-        storage_updates,
-        accessed_addresses,
-    }
+    Ok(FormattedStateUpdate { class_hash_to_compiled_class_hash, compiled_classes })
 }
 
-/// Format StateDiff's NonceUpdate to a HashMap<contract_address, nonce>
-fn format_state_update_nonces(
-    state_diff: &StateDiff,
-    transactions: &[InternalTransaction],
-) -> HashMap<Felt252, Felt252> {
-    let address_to_nonce: HashMap<Felt252, Felt252> = state_diff
-        .nonces
-        .iter()
-        .map(|nonce_update| {
-            // derive original nonce
-            // TODO: understand what is going on here XD
-            let num_nonce_bumps =
-                Felt252::from(transactions.iter().fold(0, |acc, tx| {
-                    acc + if tx.sender_address == Some(nonce_update.contract_address) { 1 } else { 0 }
-                }));
-            assert!(nonce_update.nonce > num_nonce_bumps);
-            let previous_nonce = nonce_update.nonce - num_nonce_bumps;
-            (nonce_update.contract_address, previous_nonce)
-        })
-        .collect();
-    address_to_nonce
+async fn build_compiled_class_and_maybe_update_class_hash_to_compiled_class_hash(
+    provider: &JsonRpcClient<HttpTransport>,
+    block_id: BlockId,
+    accessed_addresses: &HashSet<Felt252>,
+    address_to_class_hash: &HashMap<Felt252, Felt252>,
+    class_hash_to_compiled_class_hash: &mut HashMap<Felt252, Felt252>,
+) -> Result<HashMap<Felt252, GenericCasmContractClass>, Error> {
+    let mut compiled_classes: HashMap<Felt252, GenericCasmContractClass> = HashMap::new();
+    for contract_address in accessed_addresses {
+        let class_hash = match address_to_class_hash.get(contract_address) {
+            Some(class_hash) => class_hash,
+            None => &provider.get_class_hash_at(block_id, contract_address).await?,
+        };
+
+        let contract_class = provider.get_class(block_id, class_hash).await?;
+        let generic_sierra_cc = match contract_class {
+            starknet::core::types::ContractClass::Sierra(flattened_sierra_cc) => {
+                GenericSierraContractClass::from(flattened_sierra_cc)
+            }
+            starknet::core::types::ContractClass::Legacy(_) => {
+                unimplemented!("Fixme: Support legacy contract class")
```

Comment:
There's no Sierra notion of Sierra class for Cairo0. The compiled class is the contract class. I think the cleanest solution here will be to return a tuple of (compiled_contract_classes, deprecated_compiled_contract_classes).

Reply:
This needs to be resolved before merging the PR.